### PR TITLE
Link multiplier with org

### LINKED ISSUE #412: Link multiplier with organization
We've got a test organization for QA testing: https://github.com/ubiquibot

We also have a real organization where the bot is used: https://github.com/ubiquity

If you [set a multiplier](https://github.com/ubiquibot/production/issues/1#issuecomment-1594970863) in a QA organization (or create your own org, add "ubiquibot" and set multiplier in your own org where you are the admin) then [it will also be updated](https://github.com/ubiquity/pay.ubq.fi/issues/75#issuecomment-1594971139) in a "production" organization.

We should link a bounty hunter's multiplier to an organization where the multiplier is set.

### PR DESCRIPTION
**Pull request by devpanther**
Resolves #412

### Test

Repo1: https://github.com/one-two-test/one/issues/1#issuecomment-1655514951
Repo2: https://github.com/two-two-test/second/issues/1#issuecomment-1655514525

P.S: Org ID was used and its attached to the user_name to create a unique id

### CODE REVIEW

**Comment by devpanther** [comment]
@rndquu @pavlovcik Ready to review

**Review by rndquu — COMMENTED** [review]
While this PR solves the issue the DB schema fails to follow normal forms. I mean this `username_org_id` field.

It should have been implemented the other way:
### wallet
| user_name  | wallet_address | created_at  | updated_at |
| ------------- | ------------- | ------------- | ------------- |
### organization
| id  | github_id | name  |
| ------------- | ------------- | ------------- |
### multiplier
| id  | wallet_id | org_id  | multiplier | reason |
| ------------- | ------------- | ------------- | ------------- | ------------- |

Perhaps we can refactor it later. @0xcodercrane help

**Comment by devpanther** [comment]
> It should have been implemented the other way:

I think the wallet address should be by organization too.. and I like this schema better.

We can easily do things like listing organizations using the bot while making use of linking dynamically

**Comment by devpanther** [comment]
We should have a PR to make this change, it's going to need the organizations inserted manually once and then afterwards added automatically on install or when an event is triggered

**Comment by devpanther** [comment]
It'd also affect the `issues` and `access` because they have a field for the repository including the pending `permit` storage

**Review by rndquu — DISMISSED** [review]
@0xcodercrane @Draeieg 

Guys, we should make a priority of reviewing, QAing and rolling to production the current PR because it solves a security related issue

Thanks in advance

**Comment by pavlovcik** [comment]
> because it solves a security related issue

The security issue isn't clear to me.

**Comment by devpanther** [comment]
> > because it solves a security related issue
> 
> The security issue isn't clear to me. 

Check this https://github.com/ubiquity/pay.ubq.fi/issues/95

**Comment by rndquu** [comment]
> > because it solves a security related issue
> 
> The security issue isn't clear to me.

TLDR:
1. Bounty hunter creates **his own organization** and installs the ubiquibot
2. Bounty hunter sets multiplier to 100 in **his own organization**
3. Multiplier applies to the bounty hunter's account in the **ubiquity organization**
4. Bounty hunter self assigns an issue in the ubiquity organization and gets 100x potential payout ([example](https://github.com/ubiquity/pay.ubq.fi/issues/95#issuecomment-1655391572), thanks to @devpanther responsible disclosure we didn't miss that issue with 100x reward)

**Comment by rndquu** [comment]
@devpanther Could you resolve conflicts?

**Comment by devpanther** [comment]
> @devpanther Could you resolve conflicts?

Done

File: src/handlers/comment/handlers/assign.ts [hunk 1]
```diff
@@ -13,17 +13,23 @@ export const assign = async (body: string) => {
...
+  const organization = payload.organization;
+
   logger.info(`Received '/assign' command from user: ${payload.sender.login}, body: ${body}`);
   const issue = (_payload as Payload).issue;
   if (!issue) {
     logger.info(`Skipping '/assign' because of no issue instance`);
     return "Skipping '/assign' because of no issue instance";
   }
 
+  if (!organization?.id) {
```
Comment (rndquu) [inline src/handlers/comment/handlers/assign.ts:25]: @pavlovcik This PR basically disables the bot for none-org repos, is it ok?

File: src/handlers/comment/handlers/assign.ts [hunk 2]
```diff
@@ -13,17 +13,23 @@ export const assign = async (body: string) => {
...
+    logger.info(`Skipping '/assign' because there's no organization details`);
+    return "Skipping '/assign' because of no issue instance";
```
Comment (rndquu) [inline src/handlers/comment/handlers/assign.ts:27]: Pls make the error message more meaningful , like in the log message above: `Skipping '/assign' because there's no organization details `

File: src/handlers/comment/handlers/assign.ts (hunk 1 above)
Comment (devpanther) [reply src/handlers/comment/handlers/assign.ts:25]: Yea, i thought about that.. And looking at the marketplace, it only allowed me install it to my organization.

So I guess it works unless @pavlovcik states otherwise

File: src/handlers/comment/handlers/multiplier.ts [hunk 3]
```diff
@@ -18,6 +19,11 @@ export const multiplier = async (body: string) => {
     return;
   }
 
+  if (!organization?.id) {
+    logger.info(`Skipping '/multiplier' because there's no organization details`);
+    return;
```
Comment (rndquu) [inline src/handlers/comment/handlers/multiplier.ts:24]: Pls make sure that the bot returns a proper error message, there should not be cases when a bounty hunter executes a command and gets a silent response with no action

**Review by rndquu — CHANGES_REQUESTED** [review]

File: src/handlers/comment/handlers/multiplier.ts (hunk 3 above)
Comment (devpanther) [reply src/handlers/comment/handlers/multiplier.ts:24]: > Pls make sure that the bot returns a proper error messages for all commands where `organization.id` is required, there should not be cases when a bounty hunter executes a command and gets a silent response with no action

What about I just use `repo id` if `org id` is not available?

Comment (devpanther) [reply src/handlers/comment/handlers/multiplier.ts:24]: > Pls make sure that the bot returns a proper error messages for all commands where `organization.id` is required, there should not be cases when a bounty hunter executes a command and gets a silent response with no action

What about I just use `repo id` if `org id` is not available? @rndquu

**Comment by devpanther** [comment]
Followign up @rndquu https://github.com/ubiquity/ubiquibot/pull/557#discussion_r1278619469

File: src/handlers/comment/handlers/multiplier.ts (hunk 3 above)
Comment (rndquu) [reply src/handlers/comment/handlers/multiplier.ts:24]: But we need organization id [later in the code](https://github.com/ubiquity/ubiquibot/pull/557/files#diff-6bcca3a69bf60fc9fbb11f0b546fe3dea32062d9b5ea8672f710e06338bc6d85R77) so, as far as I understand, using repository id doesn't really make sense

Comment (devpanther) [reply src/handlers/comment/handlers/multiplier.ts:24]: > But we need organization id [later in the code](https://github.com/ubiquity/ubiquibot/pull/557/files#diff-6bcca3a69bf60fc9fbb11f0b546fe3dea32062d9b5ea8672f710e06338bc6d85R77) so, as far as I understand, using repository id doesn't really make sense

We save with either org or repo and read with same id as well.. depending on what is available

If that does not work then this will end up as an organization-only bot and bounty hunters should create organizations to test

Comment (rndquu) [reply src/handlers/comment/handlers/multiplier.ts:24]: If we use `user_id_repository_id` then multipliers are going to be repository specific so bounty hunters may end up having different multipliers for different repositories in a single organization, right?

Comment (devpanther) [reply src/handlers/comment/handlers/multiplier.ts:24]: no, its gonna be like `const id = organization?.id || repository?.id`

so, its prioritizing organizations but falling back to repository for bounty hunters test

Comment (rndquu) [reply src/handlers/comment/handlers/multiplier.ts:24]: I think it's better to keep the code consistent, I mean using `organization.id` for all cases

The original change request is regarding the error messages, so pls update them accordingly

Comment (devpanther) [reply src/handlers/comment/handlers/multiplier.ts]: > I think it's better to keep the code consistent, I mean using `organization.id` for all cases
> 
> The original change request is regarding the error messages, so pls update them accordingly

Done

File: src/handlers/comment/handlers/multiplier.ts [hunk 4]
```diff
@@ -9,13 +9,19 @@ export const multiplier = async (body: string) => {
...
   const issue = payload.issue;
   if (!issue) {
     logger.info(`Skipping '/multiplier' because of no issue instance`);
-    return;
+    return `Skipping '/multiplier' because of no issue instance`;
+  }
+
+  if (!organization?.id) {
+    logger.info(`Skipping '/assign' because the bot is not running on an organizational repository`);
+    return "Skipping '/assign' because the bot is not running on an organizational repository";
```
Comment (rndquu) [inline src/handlers/comment/handlers/multiplier.ts]:
```suggestion
    return "Skipping '/multiplier' because the bot is not running on an organizational repository";
```

File: src/handlers/comment/handlers/multiplier.ts [hunk 5]
```diff
@@ -9,13 +9,19 @@ export const multiplier = async (body: string) => {
...
 
   const issue = payload.issue;
   if (!issue) {
     logger.info(`Skipping '/multiplier' because of no issue instance`);
-    return;
+    return `Skipping '/multiplier' because of no issue instance`;
+  }
+
+  if (!organization?.id) {
+    logger.info(`Skipping '/assign' because the bot is not running on an organizational repository`);
```
Comment (rndquu) [inline src/handlers/comment/handlers/multiplier.ts]:
```suggestion
    logger.info(`Skipping '/multiplier' because the bot is not running on an organizational repository`);
```

**Review by rndquu — DISMISSED** [review]

File: src/handlers/comment/handlers/assign.ts (hunk 1 above)
Comment (0xcodercrane) [reply src/handlers/comment/handlers/assign.ts:25]: Before merging the PR. lets clarify about the installation target. 
The ubiquibot is allowed to install to org repos and individual repos so I suggest to use default value 1 if organization id is null.

**Review by 0xcodercrane — CHANGES_REQUESTED** [review]
Others are looking good.

**Comment by devpanther** [comment]
> The ubiquibot is allowed to install to org repos and individual repos

@0xcodercrane using a default value of 1 would mean all individual repo would share same multiplier value..

That is the security concern we are addressing here.

Instead of going back and fault, I can just create a `repository.id` fallback when `organization.id` is null..

**Comment by 0xcodercrane** [comment]
> Instead of going back and fault, I can just create a `repository.id` fallback when `organization.id` is null.

`repository.id` would be same as the user name?

**Comment by devpanther** [comment]
> > Instead of going back and fault, I can just create a `repository.id` fallback when `organization.id` is null.
> 
> `repository.id` would be same as the user name?

No, it'd be `${user_name}_${repository.id}`

**Comment by 0xcodercrane** [comment]
Lets go with ${user_name}_${repository.id}

**Comment by devpanther** [comment]
> Lets go with ${user_name}_${repository.id}

Done

**Review by 0xcodercrane — APPROVED** [review]
